Introduce prettyprinting to the compiler

ShinKage opened the pull request:
This patch converts most of the compiler output to the prettyprinting library in contrib.

- Pretty instances for ASTs and error messages;
- Support for colored output (enabled by default, disabled for tests);
- Color on/off option and console width option both as command line arguments and repl commands;
- Added primitive to the compiler parser for more precise text boundaries, thus more precise error messages;
- Added resolution suggestion for some error messages.
- Automatic detection of console width

This initial effort does not include proper syntax highlightning in the REPL (à la Idris1 REPL) and prettyprinting for generative commands (e.g. case split, generate definition, ...), this features can be implemented in a later PR.
At the moment, i cannot properly test these changes on platforms beside Mac, it would be great to get some feedback on other platforms, also on the format of the messages.

Note: more precise boundaries means that originally tokens carried only the position where they started, and `FC` where computed by getting this position for the starting token and the first token outside the context of the `FC`. This leads to boundaries that included whitespaces and newlines after the last token, this was particularly evident in blocks terminated by indentation (e.g. data or interface declarations). In a previous patch I've added also the start position of tokens but how it was handled in the parser was more or less an hack. In this patch the parser requires that the tokens carry both limits and thus a primitive can be introduced to avoid rewriting all the parser code. The current syntax requires a bit of nested do blocks, but hopefully will be improved. (See the no type declaration in test `idris2/reg003` for an example).

melted commented:
This will have to wait until after the 0.2.1 release as it will break the bootstrap from Idris 1.

gallais commented:
I think @edwinb wants to avoid using stuff from `contrib` in the compiler itself.

ShinKage commented:
> I think @edwinb wants to avoid using stuff from `contrib` in the compiler itself.

I thought it was fine since the compiler already depends on both contrib and network (i.e. `Control.Delayed`). In either case, we could adopt the same treatment of the parser and use a distinct version in the source tree, decoupled from the contrib version.

edwinb commented:
I prefer to avoid `contrib` but it's not an absolute rule (we do use some things already after all). The main reason for this is that `contrib` is prone to change occasionally, and I don't want to break Idris 2 itself if that happens, since everything depends on it. In some other places (notably parsing) I got around this by adding the parser libraries to the Idris 2 repo.

Anyway, we should certainly merge this once I've done the 0.2.1 release. Thanks for the contribution!

ShinKage commented:
I've added some more features and rebased to master, and I will keep it rebased until we merge without adding anymore features. I already feel that this in on the verge of too much stuff for a single PR.

The pretty printing infrastructure is there, at least for the REPL (more work is needed for IDE stuff like the underlying case splitting implementation). In future PRs, implementations of syntax highlightning or docs generation can simply add more annotations and other rendering engines while most of the work for producing the pretty printing instances for the AST is already done. Also I'm not a master printer, so if anyone has suggestion on how to improve pretty printing output, feel free to point where to improve.

NOTE: since coloring and auto-detection of console output is enabled by default, some editor can have issues with these features, nonetheless these features can be simply disabled with the newly added options in this PR. Upon merging, I will open a PR on the vim mode to disable coloring by default, since it does not support ansi escape codes without plugins or other hacks.

gallais commented:
The release has been done and this is indeed quite a big PR so I think
we can go ahead and merge it so that you don't have to keep dealing with
a moving target.
